### PR TITLE
Nested subshells

### LINKED ISSUE #198: Support nested subshells
The current parsing rules for subshells are a little naive: the subshell command will be everything from the `$(` to the first `)`, regardless of any other opening parentheses in between.

This prevent nested subshells, and also subshells containing parentheses for any other reason.

### PR DESCRIPTION
**Pull request by georgebrock**
Update the parsing rules to support nested subshells.

Fixes #198

### CODE REVIEW

File: lib/gitsh/parser.rb [hunk 1]
```diff
@@ -115,7 +115,14 @@ def parse_and_transform(command)
...
 
     rule(:subshell) do
-      str('$(') >> (str(')').absent? >> any).repeat(0).as(:subshell) >> str(')')
+      str('$(') >> (subshell_content).as(:subshell) >> str(')')
+    end
+
+    rule(:subshell_content) do
+      (
+        (str('(') >> subshell_content >> str(')')) |
+        (str(')').absent? >> any)
```
Comment (gabebw) [inline lib/gitsh/parser.rb]: What's this do?

File: spec/integration/subshell_spec.rb [hunk 2]
```diff
@@ -21,4 +21,14 @@
       expect(gitsh).to output /x in parent shell/
     end
   end
+
+  it 'supports nested subshells' do
+    GitshRunner.interactive do |gitsh|
```
Comment (gabebw) [inline spec/integration/subshell_spec.rb]: What a lovely abstraction.

**Comment by JoelQ** [comment]
LGTM, although I'm not familiar with the syntax for the parser. :shipit:

File: lib/gitsh/parser.rb (hunk 1 above)
Comment (georgebrock) [inline lib/gitsh/parser.rb]: This line is how you say “any single character except for `)`” in the Parslet DSL.

**Comment by georgebrock** [comment]
Thanks for the reviews.

**Pull request merged**